### ISSUE TITLE
Staging

### ISSUE THREAD

**Issue by mediavariance** [GIANTCRAB/gitlabby-dockerish-laravel#12]
Hi,
I'm extremely new to CD/CI, but I'm trying to wrap my head around this. I've got my gitlab-ci script successfully running the build and test jobs, but the "staging deployment" example doesn't make sense to me. 

The example staging script either pulls or pushes the repo, but does not run any other commands, including composer update/install, php artisan migrate, etc. Is this intentional? Is it a bad practice to use the docker container from the build or test jobs and push it to a staging server? Seems abnormal to have to essentially build the laravel application one more time after it's been tested successfully.

Any insight or best practices are greatly appreciated!

**Comment by GIANTCRAB** [comment]
Hi, good points stated. 

>Is it a bad practice to use the docker container from the build or test jobs and push it to a staging server?

Unless the test scripts mutate the files in the Docker container, it is OK.

>Seems abnormal to have to essentially build the laravel application one more time after it's been tested successfully.

Yes it is for most scenarios.

These scripts were written in mind about testing (this video explains why I created this - https://www.youtube.com/watch?v=Gjk0sNw3FMg ) and I admittedly add the deployment part in a hurry (which I never used). It is actually possible to just turn it into a Docker built image artefact and send it over to the endpoints. However, in that scenario, you would typically use something like Openshift/Kubernetes or Docker Swarm or Mesos to orchestrate the containers.

TL:DR; There needs to be more work on the deployment part, which I will think through more thoroughly.

**Comment by Zen0x7** [comment]
I was trying to improve the repository until I was stuck in the deployment and based on R & D I was able to understand the deployment process.

Previously i use the deployments of Laravel Envoyer. Which you just define credentials and they do:

- Create a directory with the /releases
- Pull the version to be deployed into /releases/{dateofrelease}
- Creation and modification of the symbolic link of the last deployed folder /releases/{dateofrelease} to /current
- Deployment hooks like composer install / migrate / php-fpm restart, etc.

Answering your query, when doing a pull-deploy, in my case, there is the artisan migrate, artisan install, etc.

```
# Pull Deploy
if [ "$CI_BUILD_REF_NAME" = "master" ]; then
    ssh "$STAGING_SERVER_HOSTNAME" "cd $STAGING_SERVER_DIRECTORY; git reset --hard HEAD; git checkout $GIT_DEPLOYMENT_BRANCH; git pull $GIT_DEPLOYMENT_REMOTE; composer install; php artisan optimize; php artisan migrate --force"
fi
```
But when a push deploy is performed, this is that the Git endpoint must be a bare repository. This will allow you to add hooks in the repository so that once you push it, it will be able to carry out the deployment tasks.

[Setting up receive hook for bare repo](https://stackoverflow.com/questions/14453411/setting-up-post-receive-hook-for-bare-repo)

As @GIANTCRAB  says, tests usually generate files, in my case they create compiled views. The idea is to test that the changes comply with the test plan. And once accepted, take them to a production server. It is not to copy the files from the container and move them to another server.

This weekend i'll finish my improvements of my fork and obviusly merge to this repository.

https://github.com/demency/gitlabby-dockerish-laravel

**Comment by GIANTCRAB** [comment]
After some thoughts, here's a proposed way of pushing the Docker image: 

1. Do composer installation of dependencies, etc
2. Run docker build to save the state of the image with installed dependencies
3. Run tests
4. Push image to the specified registry

For the last step, it can be a Kubernetes, Openshift, AWS, Docker, etc registry that the user must specify. How does this sound?

**Comment by GIANTCRAB** [comment]
@mediavariance I've added a Cloudfoundry version. You can use it both in your GitLab build/test/deployment process AND also in CF itself as a Docker image since it comes with Apache.

This achieves full immutable infrastructure.
